### ISSUE TITLE
[v0.5.4] error: static assertion failed: comparison object must be invocable as const

### ISSUE THREAD

**Issue by barracuda156** [capnproto/capnproto#1714]
(Not sure if this branch is still maintained, please feel free to close the issue if not.)

Since I get failures with 0.10.4, I decided to try an earlier version. It fails, though differently:
```
In file included from /opt/local/include/gcc12/c++/map:60,
                 from src/kj/main.c++:25:
/opt/local/include/gcc12/c++/bits/stl_tree.h: In instantiation of 'static const _Key& std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::_S_key(_Const_Link_type) [with _Key = const kj::MainBuilder::Impl::Option*; _Val = const kj::MainBuilder::Impl::Option*; _KeyOfValue = std::_Identity<const kj::MainBuilder::Impl::Option*>; _Compare = kj::MainBuilder::Impl::OptionDisplayOrder; _Alloc = std::allocator<const kj::MainBuilder::Impl::Option*>; _Const_Link_type = const std::_Rb_tree_node<const kj::MainBuilder::Impl::Option*>*]':
/opt/local/include/gcc12/c++/bits/stl_tree.h:2119:47:   required from 'std::pair<std::_Rb_tree_node_base*, std::_Rb_tree_node_base*> std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::_M_get_insert_unique_pos(const key_type&) [with _Key = const kj::MainBuilder::Impl::Option*; _Val = const kj::MainBuilder::Impl::Option*; _KeyOfValue = std::_Identity<const kj::MainBuilder::Impl::Option*>; _Compare = kj::MainBuilder::Impl::OptionDisplayOrder; _Alloc = std::allocator<const kj::MainBuilder::Impl::Option*>; key_type = const kj::MainBuilder::Impl::Option*]'
/opt/local/include/gcc12/c++/bits/stl_tree.h:2172:4:   required from 'std::pair<std::_Rb_tree_iterator<_Val>, bool> std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::_M_insert_unique(_Arg&&) [with _Arg = const kj::MainBuilder::Impl::Option* const&; _Key = const kj::MainBuilder::Impl::Option*; _Val = const kj::MainBuilder::Impl::Option*; _KeyOfValue = std::_Identity<const kj::MainBuilder::Impl::Option*>; _Compare = kj::MainBuilder::Impl::OptionDisplayOrder; _Alloc = std::allocator<const kj::MainBuilder::Impl::Option*>]'
/opt/local/include/gcc12/c++/bits/stl_set.h:512:25:   required from 'std::pair<typename std::_Rb_tree<_Key, _Key, std::_Identity<_Tp>, _Compare, typename __gnu_cxx::__alloc_traits<_Alloc>::rebind<_Key>::other>::const_iterator, bool> std::set<_Key, _Compare, _Alloc>::insert(const value_type&) [with _Key = const kj::MainBuilder::Impl::Option*; _Compare = kj::MainBuilder::Impl::OptionDisplayOrder; _Alloc = std::allocator<const kj::MainBuilder::Impl::Option*>; typename std::_Rb_tree<_Key, _Key, std::_Identity<_Tp>, _Compare, typename __gnu_cxx::__alloc_traits<_Alloc>::rebind<_Key>::other>::const_iterator = std::_Rb_tree<const kj::MainBuilder::Impl::Option*, const kj::MainBuilder::Impl::Option*, std::_Identity<const kj::MainBuilder::Impl::Option*>, kj::MainBuilder::Impl::OptionDisplayOrder, std::allocator<const kj::MainBuilder::Impl::Option*> >::const_iterator; typename __gnu_cxx::__alloc_traits<_Alloc>::rebind<_Key>::other = std::allocator<const kj::MainBuilder::Impl::Option*>; typename __gnu_cxx::__alloc_traits<_Alloc>::rebind<_Key> = __gnu_cxx::__alloc_traits<std::allocator<const kj::MainBuilder::Impl::Option*>, const kj::MainBuilder::Impl::Option*>::rebind<const kj::MainBuilder::Impl::Option*>; typename _Alloc::value_type = const kj::MainBuilder::Impl::Option*; value_type = const kj::MainBuilder::Impl::Option*]'
src/kj/main.c++:685:25:   required from here
/opt/local/include/gcc12/c++/bits/stl_tree.h:772:15: error: static assertion failed: comparison object must be invocable as const
  772 |               is_invocable_v<const _Compare&, const _Key&, const _Key&>,
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/local/include/gcc12/c++/bits/stl_tree.h:772:15: note: 'std::is_invocable_v<const kj::MainBuilder::Impl::OptionDisplayOrder&, const kj::MainBuilder::Impl::Option* const&, const kj::MainBuilder::Impl::Option* const&>' evaluates to false
libtool: compile:  /opt/local/bin/g++-mp-12 -DHAVE_CONFIG_H -I. -isystem/opt/local/include/LegacySupport -isystem/opt/local/include -I./src -I./src -DKJ_HEADER_WARNINGS -DCAPNP_HEADER_WARNINGS -DCAPNP_INCLUDE_DIR=\"/opt/local/include\" -D_THREAD_SAFE -pipe -Os -D_GLIBCXX_USE_CXX11_ABI=0 -arch ppc -D_THREAD_SAFE -MT src/kj/parse/char.lo -MD -MP -MF src/kj/parse/.deps/char.Tpo -c src/kj/parse/char.c++ -o src/kj/parse/char.o >/dev/null 2>&1
libtool: compile:  /opt/local/bin/g++-mp-12 -DHAVE_CONFIG_H -I. -isystem/opt/local/include/LegacySupport -isystem/opt/local/include -I./src -I./src -DKJ_HEADER_WARNINGS -DCAPNP_HEADER_WARNINGS -DCAPNP_INCLUDE_DIR=\"/opt/local/include\" -D_THREAD_SAFE -pipe -Os -D_GLIBCXX_USE_CXX11_ABI=0 -arch ppc -D_THREAD_SAFE -MT src/capnp/schema-loader.lo -MD -MP -MF src/capnp/.deps/schema-loader.Tpo -c src/capnp/schema-loader.c++ -o src/capnp/schema-loader.o >/dev/null 2>&1
make: *** [src/kj/main.lo] Error 1
```

**Comment by kentonv** [comment]
Sorry, no, we don't actively maintain old versions unless there's a security issue and that version is used by some major distro or something.

0.5.x in particular is 8 years old.

**Comment by barracuda156** [comment]
Got it, no issues here. I also rather spend effort to fix 0.10.4 and not the archaic one.
